dna/codon.go

```diff
@@ -158,107 +172,233 @@ func AminoAcidToString(a AminoAcid) string {
+	case 'C':
+		return Cys
+	case 'Q':
+		return Gln
+	case 'E':
+		return Glu
+	case 'G':
+		return Gly
+	case 'H':
+		return His
+	case 'I':
+		return Ile
+	case 'L':
+		return Leu
+	case 'K':
+		return Lys
+	case 'M':
+		return Met
+	case 'F':
+		return Phe
+	case 'P':
+		return Pro
+	case 'S':
+		return Ser
+	case 'T':
+		return Thr
+	case 'W':
+		return Trp
+	case 'Y':
+		return Tyr
+	case 'V':
+		return Val
+	case '*':
+		return Stop
+	default:
+		log.Panicf("cannot convert unrecognized byte %v to type AminoAcid", b)
+		return Stop
 	}
-	return answer
 }
 
-//CodonsToSeq reverts a slice of Codons into a slice of DNA bases.
-func CodonsToSeq(c []*Codon) []Base {
-	var answer []Base
-	for i := 0; i < len(c); i++ {
-		for j := 0; j < len(c[i].Seq); j++ {
-			answer = append(answer, c[i].Seq[j])
-		}
+// ThreeLetterToAminoAcid converts a three letter amino acid string into an AminoAcid type
+func ThreeLetterToAminoAcid(s string) AminoAcid {
+	switch s {
+	case "Ala":
+		return Ala
+	case "Arg":
+		return Arg
+	case "Asn":
+		return Asn
+	case "Asp":
+		return Asp
+	case "Cys":
+		return Cys
+	case "Gln":
+		return Gln
+	case "Glu":
+		return Glu
+	case "Gly":
+		return Gly
+	case "His":
+		return His
+	case "Ile":
+		return Ile
+	case "Leu":
+		return Leu
+	case "Lys":
+		return Lys
+	case "Met":
+		return Met
+	case "Phe":
+		return Phe
+	case "Pro":
+		return Pro
+	case "Ser":
+		return Ser
+	case "Thr":
+		return Thr
+	case "Trp":
+		return Trp
+	case "Tyr":
+		return Tyr
+	case "Val":
+		return Val
+	case "Ter":
+		return Stop
+	default:
+		log.Panicf("cannot convert unrecognized string %s into type AminoAcid", s)
+		return Stop
+	}
+}
+
+// BasesToCodons converts a slice of DNA bases into a slice of Codons.
+// Input expects bases to be in-frame. If the input sequence is not a
+// factor of three the function will panic.
+func BasesToCodons(b []Base) []Codon {
+	frame := len(b) % 3
+	answer := make([]Codon, 0, len(b)/3)
+	if !IsSeqOfACGT(b) {
+		log.Panicf("unrecognized base in sequence: %v, all input bases must be uppercase. N's are not allowed", b)
+	}
+	if frame != 0 {
+		log.Panicf("input sequence length is %d. input length must be a factor of 3", len(b))
+	}
+	for i := 0; i < len(b)-frame; i += 3 {
+		answer = append(answer, Codon{b[i], b[i+1], b[i+2]})
 	}
 	return answer
 }
 
-//TranslateCodon converts an individual Codon struct into the corresponding AminoAcid type.
-func TranslateCodon(c *Codon) AminoAcid {
-	val, ok := GeneticCode[BasesToString(c.Seq)]
-	if ok {
-		return val
-	} else {
-		log.Fatalf("Codon not in map.\n")
-		return val
+// CodonsToBases converts a slice of Codons into a slice of DNA bases.
+func CodonsToBases(c []Codon) []Base {
+	answer := make([]Base, 0, len(c)*3)
+	for i := range c {
+		answer = append(answer, c[i][0], c[i][1], c[i][2])
 	}
+	return answer
+}
+
+//TranslateCodon converts an individual Codon into the corresponding AminoAcid type.
+func TranslateCodon(c Codon) AminoAcid {
+	return GeneticCode[c]
```

Comment:
Don't we need a check here to make sure that "c" was found in the map?  Right now I think if we give something like "GGN" as the codon this will return alanine with no error